File: TreeGenerator/tree.cpp
#include <iostream>
#include <random>
#include <algorithm>
#include <utility>
#include <vector>
#include "tree.hpp"
#include "entry.hpp"

std::uniform_int_distribution<int> Node::distribution(1, 4);
std::mt19937 Node::rng(std::random_device{}());


inline void Node::addParentConnection(Entry srcEntry, Node* dstRoom, Entry dstEntry) {
  auto& connection = this->links[srcEntry];
  connection.dstRoom = dstRoom;
  connection.dstEntry = dstEntry;
  connection.isParent = true;
}

inline void Node::addConnection(Entry srcEntry, Node* dstRoom, Entry dstEntry) {
  auto& connection = this->links[srcEntry];
  connection.dstRoom = dstRoom;
  connection.dstEntry = dstEntry;
  connection.isParent = false;
}

void Node::generateSubtree(int maxDepth) {
  std::vector<Node*> nodes;
  int entry {0};

  // As the root room of the tree, generates 4 rooms and 
  // creates a connection to each of those rooms. Then 
  // generates the corresponding random subtree of rooms.
  //
  // A normal room can have [0-4] connections (including)
  // its parent room connection, but the root room has
  // always 4 connections.

  for(auto& link : this->links) {
    Entry iterEntry {(Entry)entry};
    Entry childEntry {getSymetric(iterEntry)};

    std::pair<int,int> offset = getOffset(iterEntry);
    Node* child {new Node(this->x + offset.first, this->y + offset.second)};

    child->addParentConnection(childEntry, this, iterEntry);
    this->addConnection(iterEntry, child, childEntry);
    nodes.push_back(child);
    ++entry;
  }

  for(auto& link : this->links) {
    link.dstRoom->generateSubtree(1, maxDepth, nodes);
  }

  // nodes.clear();
}

void Node::generateSubtree(int depth, int maxDepth, std::vector<Node*>& nodes) {

  // A non-root room has always one connection used to connect 
  // itself with its parent. The idea here its to get a random 
  // number of unused doors (places to stablish a connection) 
  // and create connection to new rooms. 
  std::vector<Entry> unusedLinks(FREE_LINKS);
  int doors {0};

  for(int i = 0; i < LINKS; i++) {
    if(this->links[i].dstRoom == nullptr) {
      unusedLinks.push_back((Entry)i);
      ++doors;
    }
  }

  if(doors == 0) {
    // All entries are 
    // occupied.
    return;
  }

  const int randLinks {std::min(doors, distribution(rng))};
  if(randLinks == 0) { return; }
  std::shuffle(unusedLinks.begin(), unusedLinks.end(), rng);

  for (int i {0}; i < randLinks; ++i) {
    Entry iterEntry {unusedLinks[i]};
    Entry childEntry {getSymetric(iterEntry)};

    std::pair<int,int> offset = getOffset(iterEntry);
    offset.first += this->x;
    offset.second += this->y;

    Node* child {nullptr};
    for(auto& node : nodes) {
      if(node->x == offset.first && node->y == offset.second) {
        // In case we found an already exsistent node in the 
        // target coordinates, just exit.
        child = node;
        break;
      }
    }

    // TODO When child != nullptr, randomly decide if creating a link
    // to that node or just do nothing.

    if(child == nullptr) {
      // Only create a new node if there are not any other nodes 
      // at the target coordinates.
      child = new Node(offset.first, offset.second);
      nodes.push_back(child);

      child->addParentConnection(childEntry, this, iterEntry);
      this->addConnection(iterEntry, child, childEntry);

      if((depth + 1) < maxDepth){
        child->generateSubtree(depth + 1, maxDepth, nodes);
      }
    }
  }
} 

void Node::destroySubtree(void) {
  for (auto& link : this->links) {
    if (link.dstRoom != nullptr && !link.isParent) {
      link.dstRoom->destroySubtree();
      delete link.dstRoom;
      link.dstRoom = nullptr;
    }
  }
}

void Node::printSubtree(int depth) {
  std::string fmt;
  int port = 0;

  for (int i = 0; i < depth; i++)
    fmt += "\t";

  std::cout << fmt 
    << "\033[1m\033[32m<@" << (unsigned long long) this
    << ">\033[0m" << std::endl;
       
  for (auto& link : this->links) {
    if (link.dstRoom == nullptr) {
      std::cout << fmt
        << "[" << port << "] = @null"
        << std::endl;
      
    } else {
      std::cout << fmt 
        << "[" << port << "] = @"
        << (unsigned long long)link.dstRoom
        << std::endl;
    }
    ++port;
  }

  for (auto& link : this->links) {
    if (link.dstRoom != nullptr && !link.isParent) {
      link.dstRoom->printSubtree(depth + 1);
    }
  }
}

File: TreeGenerator/src/tree.zig
const std = @import("std");

const Port = enum(u2) {
    North,
    East,
    South,
    West,

    pub fn getSymmetric(self: Port) Port {
        return switch (self) {
            Port.North => Port.South,
            Port.East => Port.West,
            Port.South => Port.North,
            Port.West => Port.East,
        };
    }
};

const NodeConnection = struct {
    dstNode: ?*Node,
    dstPort: Port,
    isParent: bool,

    pub fn new(destNode: ?*Node, destPort: Port, isParent: bool) NodeConnection {
        return NodeConnection{
            .dstNode = destNode,
            .dstPort = destPort,
            .isParent = isParent,
        };
    }
};

pub const Node = struct {
    pub var allocator: std.mem.Allocator = undefined;
    pub var rng: std.Random.DefaultPrng = undefined;
    pub var count: u32 = 0;

    connections: [4]NodeConnection,
    id: u32,
    x: u32,
    y: u32,

    pub fn newNode() Node {
        count += 1;

        return Node{
            .connections = [_]NodeConnection{
                NodeConnection.new(null, undefined, false),
            } ** 4,
            .x = 0,
            .y = 0,
            .id = count,
        };
    }

    fn addParentConnection(self: *Node, srcPort: Port, dstNode: *Node, dstPort: Port) void {
        var connection = &self.connections[@intFromEnum(srcPort)];
        connection.dstNode = dstNode;
        connection.dstPort = dstPort;
        connection.isParent = true;
    }

    fn addConnection(self: *Node, srcPort: Port, dstNode: *Node, dstPort: Port) void {
        var connection = &self.connections[@intFromEnum(srcPort)];
        connection.dstNode = dstNode;
        connection.dstPort = dstPort;
        connection.isParent = false;
    }

    pub fn generateSubtree(self: *Node, maxDepth: u32) !void {
        try self.pgenerateSubtree(0, maxDepth);
        // for (self.connections) |connection| {
        //     try connection.dstNode.?.pgenerateSubtree(1, maxDepth);
        //}
    }

    fn pgenerateSubtree(self: *Node, depth: u32, maxDepth: u32) !void {
        // Get the unused ports (ports without a connection) so we can
        // generate a random number of connections over those ports.
        var unusedPorts = [_]Port{undefined} ** 4;
        var port: u8 = 0;

        for (self.connections) |connection| {
            if (connection.dstNode == null) {
                unusedPorts[port] = @enumFromInt(port);
                port += 1;
            }
        }

        // Number of connections to generate for this node. This number
        // goes in the range [0-3]. Why not 4? One connection is already
        // in use by the father node.
        //
        // Only if the number of connections to create is different to 3,
        // shuffle the unusedPorts list (its useless otherwise). This is
        // because when we are creating connections we will pick the first
        // 'randConnections' Ports from the array to create connections
        // in those selected Ports.
        const randConnections = rng.random().int(u8) % 4;
        if (randConnections == 0) {
            return;
        } else if (randConnections != 3) {
            for (0..3) |index| {
                const rand = rng.random().int(u8) % 4;
                const tmp = unusedPorts[rand];
                unusedPorts[rand] = unusedPorts[index];
                unusedPorts[index] = tmp;
            }
        }

        for (0..randConnections) |index| {
            const nodePort = unusedPorts[index];
            const childPort = nodePort.getSymmetric();

            var child = try allocator.create(Node);
            child.* = Node.newNode();

            child.addParentConnection(childPort, self, nodePort);
            self.addConnection(nodePort, child, childPort);

            if ((depth + 1) < maxDepth) {
                try child.pgenerateSubtree(depth + 1, maxDepth);
            }
        }
    }

    pub fn destroySubtree(self: *Node) void {
        for (self.connections) |connection| {
            if (connection.dstNode != null and !connection.isParent) {
                connection.dstNode.?.destroySubtree();
                allocator.destroy(connection.dstNode.?);
            }
        }
    }

    pub fn printSubtree(self: *Node) void {
        self.pprintSubtree(0);
    }

    pub fn pprintSubtree(self: *Node, depth: u32) void {
        for (self.connections, 0..) |connection, door| {

            // Print all non-parent connections, The format is
            // [srcAddr:srcPort] -> [dstAddr:dstPort].
            if (connection.dstNode != null) {
                for (0..depth) |_| {
                    std.debug.print("\t", .{});
                }

                std.debug.print("[{x:0>4}:{}] -> [{x:0>4}:{}]\n", .{
                    self.id,
                    door,
                    connection.dstNode.?.id,
                    @intFromEnum(connection.dstPort),
                });

                if (!connection.isParent)
                    connection.dstNode.?.pprintSubtree(depth + 1);
            }
        }
    }
};
